File: typescript/packages/core/tests/guide-registry.test.ts
/**
 * Unit tests for the canonical GuideRegistry (spore-core issue #9).
 *
 * Mirrors `rust/crates/spore-core/src/guide_registry.rs#tests` one-for-one
 * (17 tests) — same rules, same verdicts, parallel structure.
 */

import { describe, expect, it } from "vitest";

import { guideRegistry, SessionId } from "../src/index.js";

const { StandardGuideRegistry, GuideId, Timestamp, GuideRegistryError, newGuideQuery } =
  guideRegistry;

type Guide = guideRegistry.Guide;
type GuideUsageRecord = guideRegistry.GuideUsageRecord;
type SessionOutcome = guideRegistry.SessionOutcome;

// ── Builders ────────────────────────────────────────────────────────────────

function ts(s: string): guideRegistry.Timestamp {
  return new Timestamp(s);
}

function makeGuide(id: string, content: string): Guide {
  return {
    id: new GuideId(id),
    name: id,
    content,
    guide_type: "skill",
    domain: null,
    source: { kind: "manual" },
    status: { kind: "active" },
    created_at: ts("2026-05-16T00:00:00Z"),
    last_used: null,
    version: 1,
  };
}

function usage(gid: string, sid: string, outcome: SessionOutcome): GuideUsageRecord {
  return {
    guide_id: new GuideId(gid),
    session_id: new SessionId(sid),
    task_domain: null,
    outcome,
    recorded_at: ts("2026-05-16T00:00:00Z"),
  };
}

// ── Tests ───────────────────────────────────────────────────────────────────

describe("GuideRegistry — register validates content", () => {
  it("register_empty_content_fails", async () => {
    const r = new StandardGuideRegistry();
    const g = makeGuide("g1", "   ");
    await expect(r.register(g)).rejects.toBeInstanceOf(GuideRegistryError);
    try {
      await r.register(g);
    } catch (e) {
      expect((e as guideRegistry.GuideRegistryError).kind).toBe("validation_failed");
    }
  });
});

describe("GuideRegistry — MetaAgentProposed forces PendingReview", () => {
  it("meta_agent_source_forces_pending_review", async () => {
    const r = new StandardGuideRegistry();
    const g = makeGuide("g1", "proposed");
    g.source = { kind: "meta_agent_proposed", proposed_at: ts("2026-05-16T01:00:00Z") };
    g.status = { kind: "active" }; // caller lies; provider corrects
    await r.register(g);

    const sel = await r.select(newGuideQuery("anything"));
    expect(sel.every((x) => x.id.value !== "g1")).toBe(true);

    const peeked = r._peekGuide(new GuideId("g1"));
    expect(peeked?.status.kind).toBe("pending_review");
    if (peeked?.status.kind === "pending_review") {
      expect(peeked.status.reason.kind).toBe("automated_proposal");
    }
  });
});

describe("GuideRegistry — select filters by status/domain/type", () => {
  it("select_filters_by_status_domain_and_type", async () => {
    const r = new StandardGuideRegistry();
    const a = makeGuide("a", "rust async tokio runtime");
    a.domain = "rust";
    const b = makeGuide("b", "pytest fixtures python");
    b.domain = "python";
    b.guide_type = "convention_doc";
    const c = makeGuide("c", "deprecated content");
    c.status = { kind: "deprecated", reason: "old", at: ts("2026-05-16T00:00:00Z") };

    await r.register(a);
    await r.register(b);
    // Inject c by registering then deprecating to avoid status validation.
    const cActive = makeGuide("c", "deprecated content");
    await r.register(cActive);
    await r.deprecate(new GuideId("c"), "old");

    const res = await r.select({
      task_instruction: "rust tokio",
      domain: "rust",
      phase: null,
      guide_types: ["skill"],
    });
    expect(res).toHaveLength(1);
    expect(res[0]!.id.value).toBe("a");
  });
});

describe("GuideRegistry — select sorted by relevance", () => {
  it("select_sorted_by_relevance", async () => {
    const r = new StandardGuideRegistry();
    await r.register(makeGuide("a", "alpha beta gamma delta"));
    await r.register(makeGuide("b", "zebra"));
    await r.register(makeGuide("c", "alpha beta"));
    const res = await r.select(newGuideQuery("alpha beta"));
    expect(res[0]!.id.value).toBe("c");
    expect(res[res.length - 1]!.id.value).toBe("b");
  });
});

describe("GuideRegistry — conflict detection at registration", () => {
  it("register_detects_conflict_in_same_domain", async () => {
    const r = new StandardGuideRegistry();
    const existing = makeGuide("a", "always run tests before commit");
    existing.domain = "rust";
    await r.register(existing);

    const conflicting = makeGuide("b", "always run tests before committing");
    conflicting.domain = "rust";
    try {
      await r.register(conflicting);
      throw new Error("expected ConflictDetected");
    } catch (e) {
      expect(e).toBeInstanceOf(GuideRegistryError);
      const err = e as guideRegistry.GuideRegistryError;
      expect(err.kind).toBe("conflict_detected");
      if (err.detail.kind === "conflict_detected") {
        expect(err.detail.conflict.guide_a.value).toBe("b");
        expect(err.detail.conflict.guide_b.value).toBe("a");
      }
    }
  });

  it("no_conflict_across_domains", async () => {
    const r = new StandardGuideRegistry();
    const a = makeGuide("a", "always run tests before commit");
    a.domain = "rust";
    await r.register(a);
    const b = makeGuide("b", "always run tests before commit");
    b.domain = "python";
    await r.register(b);
  });
});

describe("GuideRegistry — record_usage", () => {
  it("record_usage_requires_known_guide", async () => {
    const r = new StandardGuideRegistry();
    try {
      await r.recordUsage(usage("nope", "s1", { kind: "success" }));
      throw new Error("expected NotFound");
    } catch (e) {
      expect(e).toBeInstanceOf(GuideRegistryError);
      expect((e as guideRegistry.GuideRegistryError).kind).toBe("not_found");
    }
  });

  it("record_usage_updates_last_used", async () => {
    const r = new StandardGuideRegistry();
    await r.register(makeGuide("a", "x"));
    const u = usage("a", "s1", { kind: "success" });
    u.recorded_at = ts("2026-06-01T00:00:00Z");
    await r.recordUsage(u);

    const hist = await r.usageHistory(new GuideId("a"));
    expect(hist).toHaveLength(1);

    const peeked = r._peekGuide(new GuideId("a"));
    expect(peeked?.last_used?.value).toBe("2026-06-01T00:00:00Z");
  });
});

describe("GuideRegistry — deprecate", () => {
  it("deprecate_sets_status_and_404s_on_missing", async () => {
    const r = new StandardGuideRegistry();
    await r.register(makeGuide("a", "x"));
    await r.deprecate(new GuideId("a"), "obsolete");
    const peeked = r._peekGuide(new GuideId("a"));
    expect(peeked?.status.kind).toBe("deprecated");
    if (peeked?.status.kind === "deprecated") {
      expect(peeked.status.reason).toBe("obsolete");
    }
    try {
      await r.deprecate(new GuideId("nope"), "x");
      throw new Error("expected NotFound");
    } catch (e) {
      expect((e as guideRegistry.GuideRegistryError).kind).toBe("not_found");
    }
  });
});

describe("GuideRegistry — promote_to_active only from PendingReview", () => {
  it("promote_to_active_only_from_pending_review", async () => {
    const r = new StandardGuideRegistry();
    await r.register(makeGuide("a", "x"));
    try {
      await r.promoteToActive(new GuideId("a"));
      throw new Error("expected ValidationFailed");
    } catch (e) {
      expect((e as guideRegistry.GuideRegistryError).kind).toBe("validation_failed");
    }
    await r.markPendingReview(new GuideId("a"), { kind: "manual_flag", note: "x" });
    await r.promoteToActive(new GuideId("a"));
    const peeked = r._peekGuide(new GuideId("a"));
    expect(peeked?.status.kind).toBe("active");
  });
});

describe("GuideRegistry — analyze_performance", () => {
  it("analyze_performance_flags_high_failure_rate", async () => {
    const r = new StandardGuideRegistry();
    r.setNow(ts("2026-05-16T01:00:00Z"));
    await r.register(makeGuide("a", "x"));
    await r.register(makeGuide("b", "y"));
    for (let i = 0; i < 3; i++) {
      await r.recordUsage(usage("a", `s${i}`, { kind: "failure", reason: "boom" }));
    }
    for (let i = 0; i < 3; i++) {
      await r.recordUsage(usage("b", `sb${i}`, { kind: "success" }));
    }
    const signals = await r.analyzePerformance(86_400, 0.5, 100);
    const dep = signals.some(
      (s) => s.kind === "guide_deprecation_recommended" && s.guide_id.value === "a",
    );
    expect(dep).toBe(true);
  });

  it("analyze_performance_emits_skill_generation_for_repeated_pattern", async () => {
    const r = new StandardGuideRegistry();
    r.setNow(ts("2026-05-16T01:00:00Z"));
    await r.register(makeGuide("a", "x"));
    for (let i = 0; i < 4; i++) {
      await r.recordUsage(
        usage("a", `s${i}`, {
          kind: "failure",
          reason: "panic: index out of bounds",
        }),
      );
    }
    const signals = await r.analyzePerformance(86_400, 999.0, 3);
    const gen = signals.some(
      (s) =>
        s.kind === "skill_generation_needed" &&
        s.pattern === "panic: index out of bounds" &&
        s.session_ids.length === 4,
    );
    expect(gen).toBe(true);
  });

  it("analyze_performance_filters_by_window", async () => {
    const r = new StandardGuideRegistry();
    r.setNow(ts("2026-05-16T00:00:00Z"));
    await r.register(makeGuide("a", "x"));
    const old = usage("a", "s0", { kind: "failure", reason: "old-pattern" });
    old.recorded_at = ts("2020-01-01T00:00:00Z");
    await r.recordUsage(old);
    const signals = await r.analyzePerformance(3600, 0.0, 1);
    const anyPattern = signals.some((s) => s.kind === "skill_generation_needed");
    expect(anyPattern).toBe(false);
  });
});

describe("GuideRegistry — usage_history filters to one guide", () => {
  it("usage_history_filters_to_one_guide", async () => {
    const r = new StandardGuideRegistry();
    await r.register(makeGuide("a", "x"));
    await r.register(makeGuide("b", "y"));
    await r.recordUsage(usage("a", "s1", { kind: "success" }));
    await r.recordUsage(usage("b", "s2", { kind: "success" }));
    const h = await r.usageHistory(new GuideId("a"));
    expect(h).toHaveLength(1);
    expect(h[0]!.guide_id.value).toBe("a");
  });
});

describe("GuideRegistry — check_conflicts external API", () => {
  it("check_conflicts_does_not_flag_identical_content", async () => {
    const r = new StandardGuideRegistry();
    await r.register(makeGuide("a", "same exact content"));
    const conflicts = await r.checkConflicts("same exact content", null);
    expect(conflicts).toEqual([]);
  });
});

describe("GuideRegistry — select empty when no guides", () => {
  it("select_empty_when_no_guides", async () => {
    const r = new StandardGuideRegistry();
    const res = await r.select(newGuideQuery("anything"));
    expect(res).toEqual([]);
  });
});

describe("GuideRegistry — date arithmetic sanity", () => {
  it("rfc3339_round_trip", async () => {
    // Validate via analyzePerformance window math: a record exactly at `now`
    // is in-window; a record before `now - window` is not. This exercises the
    // same parse/format round-trip the Rust test asserts directly.
    const r = new StandardGuideRegistry();
    r.setNow(ts("2026-05-16T12:34:56Z"));
    await r.register(makeGuide("a", "x"));
    const u1 = usage("a", "s1", { kind: "failure", reason: "p1" });
    u1.recorded_at = ts("2026-05-16T12:34:56Z");
    await r.recordUsage(u1);
    const u2 = usage("a", "s2", { kind: "failure", reason: "p2" });
    u2.recorded_at = ts("2026-05-16T12:00:00Z");
    await r.recordUsage(u2);

    // 60-second window: only u1 should be in-window.
    const signals = await r.analyzePerformance(60, 999.0, 1);
    const patterns = signals
      .filter((s) => s.kind === "skill_generation_needed")
      .map((s) => (s as { pattern: string }).pattern);
    expect(patterns).toContain("p1");
    expect(patterns).not.toContain("p2");
  });
});

File: go/spore-core/guideregistry/guideregistry_test.go
package guideregistry

import (
	"context"
	"errors"
	"testing"
	"time"

	sporecore "github.com/squirrelsoft-dev/spore-core/go/spore-core"
)

func ts(s string) Timestamp { return Timestamp(s) }

func strPtr(s string) *string { return &s }

func makeGuide(id, content string) Guide {
	return Guide{
		ID:        GuideID(id),
		Name:      id,
		Content:   content,
		GuideType: GuideTypeSkill,
		Domain:    nil,
		Source:    NewSourceManual(),
		Status:    NewStatusActive(),
		CreatedAt: ts("2026-05-16T00:00:00Z"),
		Version:   1,
	}
}

func usage(gid, sid string, outcome SessionOutcome) GuideUsageRecord {
	return GuideUsageRecord{
		GuideID:    GuideID(gid),
		SessionID:  sporecore.SessionID(sid),
		Outcome:    outcome,
		RecordedAt: ts("2026-05-16T00:00:00Z"),
	}
}

// 1) register validates content
func TestRegisterEmptyContentFails(t *testing.T) {
	r := NewStandardGuideRegistry()
	g := makeGuide("g1", "   ")
	g.Content = "   "
	_, err := r.Register(context.Background(), g)
	var e *GuideRegistryError
	if !errors.As(err, &e) || e.Kind != ErrKindValidationFailed {
		t.Fatalf("expected ValidationFailed, got %v", err)
	}
}

// 2) MetaAgentProposed forces PendingReview
func TestMetaAgentSourceForcesPendingReview(t *testing.T) {
	r := NewStandardGuideRegistry()
	g := makeGuide("g1", "proposed")
	g.Source = NewSourceMetaAgentProposed(ts("2026-05-16T01:00:00Z"))
	g.Status = NewStatusActive() // caller lies; provider corrects
	if _, err := r.Register(context.Background(), g); err != nil {
		t.Fatalf("register: %v", err)
	}
	sel, err := r.Select(context.Background(), NewGuideQuery("anything"))
	if err != nil {
		t.Fatalf("select: %v", err)
	}
	for _, sg := range sel {
		if sg.ID == "g1" {
			t.Fatalf("g1 should not appear in Active select results")
		}
	}
	r.mu.Lock()
	defer r.mu.Unlock()
	got := r.guides[GuideID("g1")]
	if got.Status.Kind != StatusKindPendingReview {
		t.Fatalf("expected PendingReview, got %s", got.Status.Kind)
	}
	if got.Status.Reason == nil || got.Status.Reason.Kind != PendingReasonAutomatedProposal {
		t.Fatalf("expected AutomatedProposal reason, got %+v", got.Status.Reason)
	}
}

// 3) select filters by status, domain, and type
func TestSelectFiltersByStatusDomainAndType(t *testing.T) {
	r := NewStandardGuideRegistry()
	a := makeGuide("a", "rust async tokio runtime")
	a.Domain = strPtr("rust")
	b := makeGuide("b", "pytest fixtures python")
	b.Domain = strPtr("python")
	b.GuideType = GuideTypeConventionDoc
	c := makeGuide("c", "deprecated content")
	c.Status = NewStatusDeprecated("old", ts("2026-05-16T00:00:00Z"))

	if _, err := r.Register(context.Background(), a); err != nil {
		t.Fatal(err)
	}
	if _, err := r.Register(context.Background(), b); err != nil {
		t.Fatal(err)
	}
	// Inject c bypassing register's conflict check.
	r.mu.Lock()
	r.guides[GuideID("c")] = c
	r.mu.Unlock()

	res, err := r.Select(context.Background(), GuideQuery{
		TaskInstruction: "rust tokio",
		Domain:          strPtr("rust"),
		GuideTypes:      []GuideType{GuideTypeSkill},
	})
	if err != nil {
		t.Fatal(err)
	}
	if len(res) != 1 {
		t.Fatalf("expected 1 result, got %d", len(res))
	}
	if res[0].ID != "a" {
		t.Fatalf("expected a, got %s", res[0].ID)
	}
}

// 4) select sorted by relevance
func TestSelectSortedByRelevance(t *testing.T) {
	r := NewStandardGuideRegistry()
	ctx := context.Background()
	if _, err := r.Register(ctx, makeGuide("a", "alpha beta gamma delta")); err != nil {
		t.Fatal(err)
	}
	if _, err := r.Register(ctx, makeGuide("b", "zebra")); err != nil {
		t.Fatal(err)
	}
	if _, err := r.Register(ctx, makeGuide("c", "alpha beta")); err != nil {
		t.Fatal(err)
	}
	res, err := r.Select(ctx, NewGuideQuery("alpha beta"))
	if err != nil {
		t.Fatal(err)
	}
	if res[0].ID != "c" {
		t.Fatalf("expected first=c, got %s", res[0].ID)
	}
	if res[len(res)-1].ID != "b" {
		t.Fatalf("expected last=b, got %s", res[len(res)-1].ID)
	}
}

// 5) conflict detection at registration
func TestRegisterDetectsConflictInSameDomain(t *testing.T) {
	r := NewStandardGuideRegistry()
	existing := makeGuide("a", "always run tests before commit")
	existing.Domain = strPtr("rust")
	if _, err := r.Register(context.Background(), existing); err != nil {
		t.Fatal(err)
	}
	conflicting := makeGuide("b", "always run tests before committing")
	conflicting.Domain = strPtr("rust")
	_, err := r.Register(context.Background(), conflicting)
	var e *GuideRegistryError
	if !errors.As(err, &e) || e.Kind != ErrKindConflictDetected {
		t.Fatalf("expected ConflictDetected, got %v", err)
	}
	if e.Conflict == nil || e.Conflict.GuideA != "b" || e.Conflict.GuideB != "a" {
		t.Fatalf("unexpected conflict: %+v", e.Conflict)
	}
}

// 6) no conflict across domains
func TestNoConflictAcrossDomains(t *testing.T) {
	r := NewStandardGuideRegistry()
	a := makeGuide("a", "always run tests before commit")
	a.Domain = strPtr("rust")
	if _, err := r.Register(context.Background(), a); err != nil {
		t.Fatal(err)
	}
	b := makeGuide("b", "always run tests before commit")
	b.Domain = strPtr("python")
	if _, err := r.Register(context.Background(), b); err != nil {
		t.Fatalf("expected no conflict across domains, got %v", err)
	}
}

// 7) record_usage requires existing guide
func TestRecordUsageRequiresKnownGuide(t *testing.T) {
	r := NewStandardGuideRegistry()
	err := r.RecordUsage(context.Background(), usage("nope", "s1", NewOutcomeSuccess()))
	var e *GuideRegistryError
	if !errors.As(err, &e) || e.Kind != ErrKindNotFound {
		t.Fatalf("expected NotFound, got %v", err)
	}
}

// 8) record_usage updates last_used
func TestRecordUsageUpdatesLastUsed(t *testing.T) {
	r := NewStandardGuideRegistry()
	ctx := context.Background()
	if _, err := r.Register(ctx, makeGuide("a", "x")); err != nil {
		t.Fatal(err)
	}
	u := usage("a", "s1", NewOutcomeSuccess())
	u.RecordedAt = ts("2026-06-01T00:00:00Z")
	if err := r.RecordUsage(ctx, u); err != nil {
		t.Fatal(err)
	}
	hist, err := r.UsageHistory(ctx, GuideID("a"))
	if err != nil {
		t.Fatal(err)
	}
	if len(hist) != 1 {
		t.Fatalf("expected 1 record, got %d", len(hist))
	}
	r.mu.Lock()
	defer r.mu.Unlock()
	g := r.guides[GuideID("a")]
	if g.LastUsed == nil || string(*g.LastUsed) != "2026-06-01T00:00:00Z" {
		t.Fatalf("unexpected last_used: %+v", g.LastUsed)
	}
}

// 9) deprecate sets status and 404s
func TestDeprecateSetsStatusAndNotFound(t *testing.T) {
	r := NewStandardGuideRegistry()
	ctx := context.Background()
	if _, err := r.Register(ctx, makeGuide("a", "x")); err != nil {
		t.Fatal(err)
	}
	if err := r.Deprecate(ctx, GuideID("a"), "obsolete"); err != nil {
		t.Fatal(err)
	}
	r.mu.Lock()
	g := r.guides[GuideID("a")]
	r.mu.Unlock()
	if g.Status.Kind != StatusKindDeprecated || g.Status.DeprecatedReason != "obsolete" {
		t.Fatalf("expected Deprecated/obsolete, got %+v", g.Status)
	}
	err := r.Deprecate(ctx, GuideID("nope"), "x")
	var e *GuideRegistryError
	if !errors.As(err, &e) || e.Kind != ErrKindNotFound {
		t.Fatalf("expected NotFound, got %v", err)
	}
}

// 10) promote_to_active only from PendingReview
func TestPromoteToActiveOnlyFromPendingReview(t *testing.T) {
	r := NewStandardGuideRegistry()
	ctx := context.Background()
	if _, err := r.Register(ctx, makeGuide("a", "x")); err != nil {
		t.Fatal(err)
	}
	err := r.PromoteToActive(ctx, GuideID("a"))
	var e *GuideRegistryError
	if !errors.As(err, &e) || e.Kind != ErrKindValidationFailed {
		t.Fatalf("expected ValidationFailed, got %v", err)
	}
	if err := r.MarkPendingReview(ctx, GuideID("a"), NewPendingReasonManualFlag("x")); err != nil {
		t.Fatal(err)
	}
	if err := r.PromoteToActive(ctx, GuideID("a")); err != nil {
		t.Fatal(err)
	}
	r.mu.Lock()
	defer r.mu.Unlock()
	if r.guides[GuideID("a")].Status.Kind != StatusKindActive {
		t.Fatalf("expected Active after promote")
	}
}

// 11) analyze_performance flags high failure rate
func TestAnalyzePerformanceFlagsHighFailureRate(t *testing.T) {
	r := NewStandardGuideRegistry()
	ctx := context.Background()
	r.SetNow(ts("2026-05-16T01:00:00Z"))
	if _, err := r.Register(ctx, makeGuide("a", "x")); err != nil {
		t.Fatal(err)
	}
	if _, err := r.Register(ctx, makeGuide("b", "y")); err != nil {
		t.Fatal(err)
	}
	for i := 0; i < 3; i++ {
		if err := r.RecordUsage(ctx, usage("a", "s"+itoa(i), NewOutcomeFailure("boom"))); err != nil {
			t.Fatal(err)
		}
	}
	for i := 0; i < 3; i++ {
		if err := r.RecordUsage(ctx, usage("b", "sb"+itoa(i), NewOutcomeSuccess())); err != nil {
			t.Fatal(err)
		}
	}
	signals := r.AnalyzePerformance(ctx, 24*time.Hour, 0.5, 100)
	found := false
	for _, s := range signals {
		if s.Kind == SignalKindGuideDeprecationRecommended && s.GuideID == "a" {
			found = true
			break
		}
	}
	if !found {
		t.Fatalf("expected GuideDeprecationRecommended for a, got %+v", signals)
	}
}

// 12) analyze_performance emits SkillGenerationNeeded for repeated pattern
func TestAnalyzePerformanceEmitsSkillGenerationForRepeatedPattern(t *testing.T) {
	r := NewStandardGuideRegistry()
	ctx := context.Background()
	r.SetNow(ts("2026-05-16T01:00:00Z"))
	if _, err := r.Register(ctx, makeGuide("a", "x")); err != nil {
		t.Fatal(err)
	}
	for i := 0; i < 4; i++ {
		if err := r.RecordUsage(ctx, usage("a", "s"+itoa(i), NewOutcomeFailure("panic: index out of bounds"))); err != nil {
			t.Fatal(err)
		}
	}
	signals := r.AnalyzePerformance(ctx, 24*time.Hour, 999.0, 3)
	found := false
	for _, s := range signals {
		if s.Kind == SignalKindSkillGenerationNeeded &&
			s.Pattern == "panic: index out of bounds" &&
			len(s.SessionIDs) == 4 {
			found = true
			break
		}
	}
	if !found {
		t.Fatalf("expected SkillGenerationNeeded, got %+v", signals)
	}
}

// 13) analyze_performance honors window
func TestAnalyzePerformanceFiltersByWindow(t *testing.T) {
	r := NewStandardGuideRegistry()
	ctx := context.Background()
	r.SetNow(ts("2026-05-16T00:00:00Z"))
	if _, err := r.Register(ctx, makeGuide("a", "x")); err != nil {
		t.Fatal(err)
	}
	old := usage("a", "s0", NewOutcomeFailure("old-pattern"))
	old.RecordedAt = ts("2020-01-01T00:00:00Z")
	if err := r.RecordUsage(ctx, old); err != nil {
		t.Fatal(err)
	}
	signals := r.AnalyzePerformance(ctx, time.Hour, 0.0, 1)
	for _, s := range signals {
		if s.Kind == SignalKindSkillGenerationNeeded {
			t.Fatalf("expected no SkillGenerationNeeded for old record, got %+v", s)
		}
	}
}

// 14) usage_history filters to one guide
func TestUsageHistoryFiltersToOneGuide(t *testing.T) {
	r := NewStandardGuideRegistry()
	ctx := context.Background()
	if _, err := r.Register(ctx, makeGuide("a", "x")); err != nil {
		t.Fatal(err)
	}
	if _, err := r.Register(ctx, makeGuide("b", "y")); err != nil {
		t.Fatal(err)
	}
	if err := r.RecordUsage(ctx, usage("a", "s1", NewOutcomeSuccess())); err != nil {
		t.Fatal(err)
	}
	if err := r.RecordUsage(ctx, usage("b", "s2", NewOutcomeSuccess())); err != nil {
		t.Fatal(err)
	}
	h, err := r.UsageHistory(ctx, GuideID("a"))
	if err != nil {
		t.Fatal(err)
	}
	if len(h) != 1 || h[0].GuideID != "a" {
		t.Fatalf("expected only a's history, got %+v", h)
	}
}

// 15) check_conflicts does not flag identical content
func TestCheckConflictsDoesNotFlagIdenticalContent(t *testing.T) {
	r := NewStandardGuideRegistry()
	ctx := context.Background()
	if _, err := r.Register(ctx, makeGuide("a", "same exact content")); err != nil {
		t.Fatal(err)
	}
	conflicts := r.CheckConflicts(ctx, "same exact content", nil)
	if len(conflicts) != 0 {
		t.Fatalf("expected no conflicts, got %+v", conflicts)
	}
}

// 16) select empty when no guides
func TestSelectEmptyWhenNoGuides(t *testing.T) {
	r := NewStandardGuideRegistry()
	res, err := r.Select(context.Background(), NewGuideQuery("anything"))
	if err != nil {
		t.Fatal(err)
	}
	if len(res) != 0 {
		t.Fatalf("expected empty, got %+v", res)
	}
}

// 17) RFC3339 round-trip sanity
func TestRFC3339RoundTrip(t *testing.T) {
	cutoff, ok := rfc3339Subtract(ts("2026-05-16T12:34:56Z"), 0)
	if !ok {
		t.Fatalf("expected parse ok")
	}
	if string(cutoff) != "2026-05-16T12:34:56Z" {
		t.Fatalf("expected round-trip, got %s", cutoff)
	}
}

// ── helpers ────────────────────────────────────────────────────────────────

func itoa(i int) string {
	// simple itoa for test session ids
	if i == 0 {
		return "0"
	}
	neg := false
	if i < 0 {
		neg = true
		i = -i
	}
	buf := make([]byte, 0, 6)
	for i > 0 {
		buf = append([]byte{byte('0' + i%10)}, buf...)
		i /= 10
	}
	if neg {
		return "-" + string(buf)
	}
	return string(buf)
}
